test/pjrt/test_collective_ops_tpu.py

```diff
@@ -132,5 +138,96 @@ def test_all_to_all(self, pin_layout):
                                              list(range(world_size))]])
 
 
+# Test for collective ops from torch.distributed
+class TestDistCollectiveOpsTpu(parameterized.TestCase):
+
+  # TODO(zpcore): fix the openxla dynamo issue for dist.all_reduce
+  @staticmethod
+  def my_compiler(gm: torch.fx.GraphModule, example_inputs: List[torch.Tensor]):
+    return gm.forward
+
+  @staticmethod
+  def _all_reduce(use_dynamo: bool):
+    met.clear_all()
+
+    def _dynamo_func_wrapper(input):
+      dist.all_reduce(input, dist.ReduceOp.SUM)
+      return input
+
+    dist.init_process_group("xla", init_method='xla://')
+    device = xm.xla_device()
+    if xm.xla_device_hw(device) not in ('TPU', 'CUDA', 'NEURON'):
+      print(f'skip this test for hw {xm.xla_device_hw(device)}')
+      return
+    ordinal = torch.tensor([xr.global_ordinal()],
+                           dtype=torch.float,
+                           device=device)
+    input = ordinal
+    if not use_dynamo:
+      dist.all_reduce(input)
+      xm.mark_step()
+      assert 'xla::AllReduceInPlace' in met.counter_names(
+      ) or 'xla::AllReduce' in met.counter_names()
+    else:
+      compiled_collective = torch.compile(
+          _dynamo_func_wrapper, backend=TestDistCollectiveOpsTpu.my_compiler)
+      compiled_collective(input)
+      torch_xla.sync()
+      assert 'xla::all_reduce' in met.counter_names()
+    return input.cpu()
+
+  @staticmethod
+  def _all_gather_into_tensor(use_dynamo: bool):
+    met.clear_all()
+
+    def _dynamo_func_wrapper(input):
+      output_tensor = torch.empty((1, xr.world_size()), device=device)
+      dist.all_gather_into_tensor(output_tensor, input, None)
+      return output_tensor
+
+    dist.init_process_group("xla", init_method='xla://')
+    device = xm.xla_device()
+    if xm.xla_device_hw(device) not in ('TPU', 'CUDA', 'NEURON'):
+      print(f'skip this test for hw {xm.xla_device_hw(device)}')
+      return
+    ordinal = torch.tensor([xr.global_ordinal()],
+                           dtype=torch.float,
+                           device=device)
+    input = ordinal
+
+    if not use_dynamo:
+      output_tensor = torch.empty((1, xr.world_size()), device=device)
+      dist.all_gather_into_tensor(output_tensor, input, None)
+      xm.mark_step()
+      assert 'xla::AllGather' in met.counter_names(
+      ) or 'xla::AllGatherOut' in met.counter_names()
+    else:
+      compiled_collective = torch.compile(
+          _dynamo_func_wrapper, backend='openxla')
+      output_tensor = compiled_collective(input)
+      torch_xla.sync()
+      assert 'xla::all_gather_into_tensor' in met.counter_names()
+    return output_tensor.cpu()
+
+  @parameterized.named_parameters(('dynamo', True), ('nondynamo', False))
+  @mock.patch.object(xr, 'world_size', return_value=tpu.num_local_processes())
```

Comment:
Why is world size mocked out?

Reply:
It's a work around to call `pjrt.run_multiprocess` multiple times due to all `@parameterized.named_parameters` wrapped test functions are launched by reusing the same process one by one.

Without the mock, this error will be triggered: https://github.com/pytorch/xla/blob/ae28308bed8ef2a2086b3d7993fce766b34daf3b/torch_xla/_internal/pjrt.py#L144-L146
If we are not calling xr.world_size(), then `_xla_get_replication_devices_count` and `global_device_count` will not be called: https://github.com/pytorch/xla/blob/ae28308bed8ef2a2086b3d7993fce766b34daf3b/torch_xla/runtime.py#L136-L145  thus we will not trigger the `std::atomic<bool> g_computation_client_initialized`.

Reply:
Just tested that `tpu.num_local_processes()` or `tpu.num_expected_global_devices()` will break the dynamo graph. I think stick to `runtime.world_size()` and use the mock object will be most straightforward way for the unit test. WDYT?

Reply:
Hmm, that doesn't sound right

`Runtime is already initialized` happens if you do `xr.world_size` in the same process where you call `launch`, so that's expected. But, you really only need it for your test assertion, which you can put with the other test assertions in `self._all_reduce`.

The only dynamo compiled function is `_dist_allgather`. Where are `tpu.num_local_processes()` or `tpu.num_expected_global_devices()` called in there? Also I'm 90% sure this mock does not apply to anything in the child process spawned by `pjrt.run_multiprocess`

Reply:
OK, I think the thumb of rule to prevent the `Runtime is already initialized...` error is:
Never initialize PJRT client (e.g., call xr.world_size()) before calling `pjrt.run_multiprocess()`.

For parameterized test that calls `pjrt.run_multiprocess()`, do not call `xr.world_size()` in the main test process. Because `pjrt.run_multiprocess()` will be triggered in the next parameter test, which violates rule.

I have removed the xr.world_size() mock object and use `tpu.num_expected_global_devices()` in the test process instead.